File: DS-OU/MergeSort.c
// take an array
#include <stdio.h>

int array[10] = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0};

// write a function to divide the array every time into 2 parts till the size of array is 1
void MergeSort(int *array, int left, int right)
{

    if (left >= right)
    {
        return 0;
    }

    int mid = (left + right) / 2;
    MergeSort(array, left, mid);
    MergeSort(array, mid + 1, right);

    int merged_sorted_array_index = 0;
    int swapper = 0;

    while (left < mid && mid + 1 <= right)
    {
        if (array[left] > array[right])
        {
            swapper = array[right];
            array[right] = array[left];
            array[left] = swapper;
            left++;
        }
        else
        {
            swapper = array[right];
            array[right] = array[left];
            array[left] = swapper;
            right++;
        }
    }
    if (left == mid)
    {
        while (mid + 1 <= right)
        {
        }
    }
}

File: DS-OU/MergeSort.py
# create an array
array = [9,8,7,6,5,4,3,2,1,0]

def MergeSort(array):
    
    if(len(array) == 1):
        return array
    # divide the array into 2 parts every time till the size of array is 1
    left = MergeSort(array[:len(array)//2])
    right = MergeSort(array[len(array)//2:])


    left_index = 0
    right_index = 0

    # creating another array to store merged and sorted array
    merged_sorted_array = []

    while(left_index < len(left) and right_index < len(right)):
        if(left[left_index] < right[right_index]):
            merged_sorted_array.append(left[left_index])
            left_index += 1
        else:
            merged_sorted_array.append(right[right_index])
            right_index += 1
    
    if(left_index == len(left)):
        while(right_index < len(right)):
            merged_sorted_array.append(right[right_index])
            right_index += 1
    else:
        while(left_index < len(left)):
            merged_sorted_array.append(left[left_index])
            left_index += 1
            
    return merged_sorted_array


print(MergeSort(array))